vllm/spec_decode/spec_decode_worker.py

```diff
@@ -199,27 +210,36 @@ def execute_model(
             "speculative decoding "
             "requires non-None seq_group_metadata_list")
 
+        # When the queue size is too large, disable speculative decoding
+        # to stop trading off throughput for latency.
+        disable_all = (execute_model_req.running_queue_size >=
+                       self.disable_at_queue_size)
+        if disable_all:
+            for seq_group_metadata in execute_model_req.seq_group_metadata_list:
+                seq_group_metadata._num_speculative_tokens = 0
```

Comment:
The purpose I currently keep this along with SequenceGroupMetadata is to let it naturally have the same lifecycle of requests. A potential issue with our internal method with `non_spec_seq_ids` (which is similar to what you proposed here) is we have to find a way to clean up finished requests from this dict; otherwise it will get bigger over time.

Another reason we need this information is our current speculative decoding implementation doesn't support "re-enable" speculative decoding for a request once it has been disabled, because the kv-cache of the draft model is not updated anymore. As a result, we need to make sure once a request has been disabled, it has to be disabled forever.

Reply:
Yeah, the way to do that is to track the state in the worker and have `ExecuteModelRequest` also store aborted/finished request ids (we do this in our fork). The issue with putting more things in the Sequence classes is it couples the whole system together, even when only one layer needs it. 

I won't block the PR on this but it's trending the wrong way in terms of software design.

Reply:
Let's at least clean up the APIs so we're not setting protected fields

Reply:
and add a comment that we store spec decode-specific state in the global datastructures

Reply:
Agree. I'll clean this up and put a TODO for future refactoring in DSD.